libr/bin/format/elf/elf.c

```diff
@@ -1277,228 +1299,185 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
-	} r;
-	bool is_rela;
-	int k;
-};
 
-static RBinElfSection *get_rel_sec(ELFOBJ *bin, const char **sects) {
-	RBinElfSection *rel_sec = NULL;
-	int j = 0;
-	while (!rel_sec && sects[j]) {
-		rel_sec = get_section_by_name (bin, sects[j++]);
-	}
-	return rel_sec;
+static void rel_cache_free(HtUPKv *kv) {
+	free (kv->value);
 }
 
-static void read_rel(ELFOBJ *bin, Elf_(Rel) *rel, ut8 *rl) {
-	int l = 0;
-	rel->r_offset = READWORD (rl, l);
-	rel->r_info = READWORD (rl, l);
-}
+static HtUP *rel_cache_new(ELFOBJ *bin) {
+	RBinElfReloc *relocs = Elf_(r_bin_elf_get_relocs) (bin);
+	const int htsize = R_MIN (bin->reloc_num, 1024);
+	HtUP *rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	size_t i;
 
-static void read_rela(ELFOBJ *bin, Elf_(Rela) *rela, ut8 *rl) {
-	int l = 0;
-	rela->r_offset = READWORD (rl, l);
-	rela->r_info = READWORD (rl, l);
-	rela->r_addend = READWORD (rl, l);
+	for (i = 0; i < bin->reloc_num; i++) {
+		RBinElfReloc *tmp = R_NEW (RBinElfReloc);
+		memcpy (tmp, relocs + i, sizeof (RBinElfReloc));
+		if (!ht_up_insert (rel_cache, tmp->sym, tmp)) {
+			free (tmp);
+		}
+	}
+	free (relocs);
+	return rel_cache;
 }
 
-static struct ht_rel_t *read_ht_rel(ELFOBJ *bin, ut8 *rl, int k) {
-	struct ht_rel_t *rel = R_NEW0 (struct ht_rel_t);
-	if (!rel) {
-		return NULL;
+static ut64 get_got_entry(ELFOBJ *bin, RBinElfReloc *rel) {
+	if (!rel->rva) {
+		return UT64_MAX;
 	}
 
-	rel->is_rela = bin->is_rela == DT_RELA;
-	rel->k = k;
-	if (rel->is_rela) {
-		read_rela (bin, &rel->r.rela, rl);
-	} else {
-		read_rel (bin, &rel->r.rel, rl);
+	ut64 p_sym_got_addr = Elf_(r_bin_elf_v2p_new) (bin, rel->rva);
+
+	ut8 buf[WORDSIZE];
+	int res = r_buf_read_at (bin->b, p_sym_got_addr, buf, WORDSIZE);
+
+	if (res != WORDSIZE) {
+		return UT64_MAX;
+	}
+
+	size_t i = 0;
+	ut64 addr = READWORD (buf, i);
+
+	if (!addr) {
+		return UT64_MAX;
 	}
-	return rel;
+
+	return addr;
 }
 
-static void rel_cache_free(HtUPKv *kv) {
-	free (kv->value);
+static bool is_thumb_symbol(ut64 plt_addr) {
+	return plt_addr & 1;
 }
 
-static HtUP *rel_cache_new(ELFOBJ *bin) {
-	ut8 rl[MAX_REL_RELA_SZ] = { 0 };
-	RBinElfSection *rel_sec = NULL;
-	int j, k, tsize, nrel;
-	const char *rel_sect[] = { ".rel.plt", ".rela.plt", ".rel.dyn", ".rela.dyn", NULL };
-	const char *rela_sect[] = { ".rela.plt", ".rel.plt", ".rela.dyn", ".rel.dyn", NULL };
-	HtUP *rel_cache;
+static ut64 get_import_addr_arm(ELFOBJ *bin, RBinElfReloc *rel) {
+	ut64 got_addr = bin->dyn_info.dt_pltgot;
+	if (!got_addr) {
+		return UT64_MAX;
+	}
 
-	if ((!bin->shdr || !bin->strtab) && !bin->phdr) {
-		return NULL;
+	ut64 plt_addr = get_got_entry (bin, rel);
+	if (plt_addr == UT64_MAX) {
+		return UT64_MAX;
 	}
 
-	if (bin->is_rela == DT_REL) {
-		rel_sec = get_rel_sec (bin, rel_sect);
-		tsize = sizeof (Elf_(Rel));
-	} else if (bin->is_rela == DT_RELA) {
-		rel_sec = get_rel_sec (bin, rela_sect);
-		tsize = sizeof (Elf_(Rela));
+	ut64 pos = (rel->rva - got_addr - 0x3 * WORDSIZE) / WORDSIZE;
```

Comment:
it's not clear what it does. 
For example, why `3 * WORDSIZE`? (I guess there are 3 fields of WORDSIZE size, but...)
Also, why didn't use you store `pos` field in the relocation entry like we had before with the `k` field?

Reply:
this probably deserves a separate function or a macro with a meaningful name (e.g. `COMPUTE_PLTGOT_POSITION (rel, pltgot_addr, n_initial_unused_entries)`) and you can reuse this in many places to make it clearer.